toolchain/lower/constant.cpp

```diff
@@ -0,0 +1,270 @@
+}
+
+static auto EmitBoundMethodAsConstant(ConstantContext& context,
+                                      SemIR::BoundMethod inst)
+    -> llvm::Constant* {
+  // Propagate just the function; the object is separately provided to the
+  // enclosing call as an implicit argument.
+  return context.GetConstant(inst.function_id);
+}
+
+static auto EmitFieldDeclAsConstant(ConstantContext& context,
+                                    SemIR::FieldDecl inst) -> llvm::Constant* {
+  return context.GetUnusedConstant(inst.type_id);
+}
+
+static auto EmitFloatLiteralAsConstant(ConstantContext& context,
+                                       SemIR::FloatLiteral inst)
+    -> llvm::Constant* {
+  const llvm::APFloat& value = context.sem_ir().floats().Get(inst.float_id);
+  return llvm::ConstantFP::get(context.GetType(inst.type_id), value);
+}
+
+static auto EmitFunctionDeclAsConstant(ConstantContext& context,
+                                       SemIR::FunctionDecl inst)
+    -> llvm::Constant* {
+  // TODO: Add a FunctionValue that FunctionDecl evaluates to, and remove this.
+  return context.GetFunction(inst.function_id);
+}
+
+static auto EmitInterfaceWitnessAsConstant(ConstantContext& context,
+                                           SemIR::InterfaceWitness inst)
+    -> llvm::Constant* {
+  // TODO: For dynamic dispatch, we might want to lower witness tables as
+  // constants.
+  return context.GetUnusedConstant(inst.type_id);
+}
+
+static auto EmitIntLiteralAsConstant(ConstantContext& context,
+                                     SemIR::IntLiteral inst)
+    -> llvm::Constant* {
+  return llvm::ConstantInt::get(context.GetType(inst.type_id),
+                                context.sem_ir().ints().Get(inst.int_id));
+}
+
+static auto EmitNamespaceAsConstant(ConstantContext& context,
+                                    SemIR::Namespace inst) -> llvm::Constant* {
+  return context.GetUnusedConstant(inst.type_id);
+}
+
+static auto EmitRealLiteralAsConstant(ConstantContext& context,
+                                      SemIR::RealLiteral inst)
+    -> llvm::Constant* {
+  const Real& real = context.sem_ir().reals().Get(inst.real_id);
+  // TODO: This will probably have overflow issues, and should be fixed.
+  double val =
+      real.mantissa.getZExtValue() *
+      std::pow((real.is_decimal ? 10 : 2), real.exponent.getSExtValue());
+  llvm::APFloat llvm_val(val);
+  return llvm::ConstantFP::get(context.GetType(inst.type_id), llvm_val);
+}
+
+static auto EmitStringLiteralAsConstant(ConstantContext& /*context*/,
+                                        SemIR::StringLiteral inst)
+    -> llvm::Constant* {
+  CARBON_FATAL() << "TODO: Add support: " << inst;
+}
+
+static auto EmitStructTypeFieldAsConstant(ConstantContext& /*context*/,
+                                          SemIR::StructTypeField /*inst*/)
+    -> llvm::Constant* {
+  // A StructTypeField isn't a value, so this constant value won't ever be used.
+  return nullptr;
```

Comment:
Why isn't this returning GetUnusedConstant?

Reply:
Added a comment to explain.